source/h2_stream.c

```diff
@@ -324,21 +345,34 @@ int aws_h2_stream_encode_data_frame(
 
     *out_has_more_data = false;
     *out_stream_stalled = false;
+    *flow_controlled = false;
+    struct aws_h2_connection *connection = s_get_h2_connection(stream);
+
+    if (stream->thread_data.stalled) {
+        /* The stream has negative window size now */
+        *out_has_more_data = true;
+        *flow_controlled = true;
+        return AWS_OP_SUCCESS;
```

Comment:
We may still need to check the stream window_size_peer in my understanding.

Since when changing of SETTING_INITIAL_WINDOW_SIZE causes the window of an open stream to 0 or negative, it may be better to let the outgoing task notify it from here than just moving them into the stalled list.  

But yeah, we should precondition on connection window size.